File: ruby/lib/todo/list.rb
module Todo
  class List

    def initialize(items)
      @items = items
      @tasks = Hash.new
      @tasks['open'] = []
      @tasks['completed'] = []
      @items.each do |item|
        if item.is_a? Task
          if item.completed?
            @tasks['completed'] << item
          else
            @tasks['open'] << item
          end
        end
      end
    end

    def items
      @items
    end

    def tasks
      @tasks['open'] + @tasks['completed']
    end

    def open
      @tasks['open']
    end

    def completed
      @tasks['completed']
    end

    def add(task)
      if task.completed?
        @tasks['completed'] << task
      else
        @tasks['open'] << task
      end
      @items << task
    end

    def find(identifier)
      pattern = Regexp.new identifier
      tasks.each do |task|
        return task if task.identifier =~ pattern
      end

      return nil
    end

    def delete(identifier)
      task = find identifier
      return "The task \"#{identifier}\" does not exist" if task.nil?

      if task.completed?
        @tasks['completed'].delete task
      else
        @tasks['open'].delete task
      end
      @items.delete task
    end

    def update(identifier, new_identifier = nil, estimated_time = nil, priority = nil)
      task = find identifier
      return "The task \"#{identifier}\" does not exist" if task.nil?

      task.update new_identifier, estimated_time, priority
    end
  end
end

File: java/src/main/java/me/ramblingsby/todos/List.java
package me.ramblingsby.todos;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class List {

  private java.util.List<Item> items;
  private java.util.List<Task> openTasks;
  private java.util.List<Task> completedTasks;

  public List(java.util.List<Item> items) {
    this.items = items;
    openTasks = new ArrayList<Task>();
    completedTasks = new ArrayList<Task>();

    for (Item item : items) {
      if (item.getClass() == Task.class) {
        Task task = (Task)item;
        if (task.completed()) {
          completedTasks.add(task);
        } else {
          openTasks.add(task);
        }
      }
    }
  }

  public java.util.List<Task> tasks() {
    java.util.List<Task> tasks = new ArrayList<Task>(openTasks);
    tasks.addAll(completedTasks);
    return tasks;
  }

  public java.util.List<Task> openTasks() {
    return new ArrayList<Task>(openTasks);
  }

  public java.util.List<Task> completedTasks() {
    return new ArrayList<Task>(completedTasks);
  }

  public java.util.List<Item> items() {
    return new ArrayList<Item>(items);
  }

  public void add(Task task) {
    if (task.completed()) {
      completedTasks.add(task);
    } else {
      openTasks.add(task);
    }
    items.add(task);
  }

  public Task find(String patternString) {
    Pattern pattern = Pattern.compile(patternString);
    for (Task task : tasks()) {
      Matcher matcher = pattern.matcher(task.getIdentifier());
      if (matcher.find()) {
        return task;
      }
    }

    return null;
  }

  public Object delete(String identifier) {
    Task task = find(identifier);
    if (task == null) {
      return "The task \"" + identifier + "\" does not exist";
    }

    if (task.completed()) {
      completedTasks.remove(task);
    } else {
      openTasks.remove(task);
    }

    return items.remove(task);
  }

  public Object update(String identifier, String newIdentifier, String estimatedTime, String priority) {
    Task task = find(identifier);
    if (task == null) {
      return "The task \"" + identifier + "\" does not exist";
    }

    return task.update(newIdentifier, estimatedTime, priority);
  }
}
